src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -351,21 +409,279 @@ ErrorIfModifyQueryNotSupported(Query *queryTree)
+ *
+ * We special-case CASE/COALESCE because those are evaluated lazily. We could evaluate
+ * CASE/COALESCE expressions which don't reference Vars, or partially evaluate some
+ * which do, but for now we just error out. That makes both the code and user-education
+ * easier.
+ */
+static bool
+ContainsDisallowedFunctionCalls(Node *expression, bool *varArgument, bool *badCoalesce)
+{
+	bool result;
+	WalkerState data;
+	data.containsVar = data.varArgument = data.badCoalesce = false;
+
+	result = ContainsDisallowedFunctionCallsWalker(expression, &data);
+
+	*varArgument |= data.varArgument;
+	*badCoalesce |= data.badCoalesce;
+	return result;
+}
+
+
+static bool
+ContainsDisallowedFunctionCallsWalker(Node *expression, WalkerState *state)
+{
+	char volatileFlag = 0;
+	WalkerState childState;
+	bool containsDisallowedFunction = false;
+
+	childState.containsVar = childState.varArgument = childState.badCoalesce = false;
+
+	if (expression == NULL)
 	{
-		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("modifying the partition value of rows is not allowed")));
+		return false;
+	}
+
+	if (IsA(expression, CoalesceExpr))
+	{
+		CoalesceExpr* expr = (CoalesceExpr *) expression;
+
+		if (contain_mutable_functions((Node *) (expr->args)))
+		{
+			state->badCoalesce = true;
+			return true;
+		}
+		else
+		{
+			/*
+			 * There's no need to recurse. Since there are no STABLE functions
+			 * varArgument will never be set.
+			 */
+			return false;
+		}
+	}
+
+	if (IsA(expression, CaseExpr))
+	{
+		CaseExpr* expr = (CaseExpr *) expression;
+		ListCell *temp;
+
+		/*
+		 * contain_mutable_functions doesn't know what to do with CaseWhen so we
+		 * have to break it out ourselves
+		 */
+		foreach(temp, expr->args)
+		{
+			CaseWhen *when = (CaseWhen *) lfirst(temp);
+			Assert(IsA(when, CaseWhen));
+
+			if (contain_mutable_functions((Node *) when->expr) ||
+				contain_mutable_functions((Node *) when->result))
+			{
+				state->badCoalesce = true;
+				return true;
+			}
+		}
+
+		if (contain_mutable_functions((Node *) expr->defresult))
+		{
+			state->badCoalesce = true;
+			return true;
+		}
+
+		return ContainsDisallowedFunctionCallsWalker((Node *) (expr->arg), state);
+	}
+
+	if (IsA(expression, Var))
+	{
+		state->containsVar = true;
+		return false;
+	}
+
+	/*
+	 * In order for statement replication to give us consistent results it's important
+	 * that we either disallow or evaluate on the master anything which has a volatility
+	 * category above IMMUTABLE. Newer versions of postgres might add node types which
+	 * should be checked in this function.
+	 *
+	 * Look through contain_mutable_functions_walker or future PG's equivalent for new
+	 * node types before bumping this version number to fix compilation.
+	 *
+	 * Once you've added them to this check, make sure you also evaluate them in the
+	 * executor!
+	 */
+	StaticAssertStmt(PG_VERSION_NUM <= 90503, "When porting to a newer PG this section"
```

Comment:
This shouldn't test for a minor version, rather for a major version. New node types aren't added in minor releases, and having to adapt this test that often would be a pain.